shell-setup/src/main.ts

```diff
@@ -0,0 +1,330 @@
+    append = command.append ?? "";
+  }
+  if (!prepend && !append) {
+    return false;
+  }
+
+  let contents: string | undefined;
+  try {
+    contents = await readTextFile(rc);
+    if (prepend) {
+      if (contents.includes(prepend)) {
+        // nothing to prepend
+        prepend = "";
+      } else {
+        // always add a newline
+        prepend = ensureEndsWith(prepend, "\n");
+      }
+    }
+    if (append) {
+      if (contents.includes(append)) {
+        // nothing to append
+        append = "";
+      } else if (!contents.endsWith("\n")) {
+        // add new line to start
+        append = ensureStartsWith(append, "\n");
+      }
+    }
+  } catch (_error) {
+    prepend = prepend ? ensureEndsWith(prepend, "\n") : prepend;
+    append = append ? ensureStartsWith(append, "\n") : append;
+  }
+  if (!prepend && !append) {
+    return false;
+  }
+
+  if (contents !== undefined) {
+    await backups.add(rc, contents);
+  }
+
+  await ensureExists(dirname(rc));
+
+  try {
+    await writeTextFile(rc, prepend + (contents ?? "") + append, {
+      create: true,
+    });
+
+    return true;
+  } catch (error) {
+    if (
+      error instanceof Deno.errors.PermissionDenied ||
+      // deno-lint-ignore no-explicit-any
+      error instanceof (Deno.errors as any).NotCapable
+    ) {
+      return false;
+    }
+    throw withContext(`Failed to update shell rc file: ${rc}`, error);
+  }
+}
+
+/** Write the commands necessary to source the env file (which sets up the path).
+ * Up until this point, we have not modified any shell config files.
+ */
+async function addToPath(
+  availableShells: UnixShell[],
+  installDir: string,
+  backups: Backups,
+) {
+  for (const shell of availableShells) {
+    const sourceCmd = await (shell.sourceString ?? shSourceString)(installDir);
+
+    for (const rc of await shell.rcsToUpdate()) {
+      await updateRcFile(rc, sourceCmd, backups);
+    }
+  }
+}
+
+// Update this when adding support for a new shell
+const shells: UnixShell[] = [
+  new Posix(),
+  new Bash(),
+  new Zsh(),
+  new Fish(),
+];
+
+async function getAvailableShells(): Promise<UnixShell[]> {
+  const present = [];
+  for (const shell of shells) {
+    if (await shell.exists()) {
+      present.push(shell);
+    }
+  }
+  return present;
+}
+
+async function setupShells(installDir: string, backupDir: string) {
+  const availableShells = await getAvailableShells();
+
+  await writeEnvFiles(availableShells, installDir);
+
+  const backups = new Backups(backupDir);
+
+  if (
+    await confirm(`Edit shell configs to add deno to the PATH?`, {
+      default: true,
+    })
+  ) {
+    await ensureExists(backupDir);
+    await addToPath(availableShells, installDir, backups);
+  }
+
+  const shellsWithCompletion = availableShells.filter((s) =>
+    s.supportsCompletion !== false
+  );
+  const selected = await multiSelect(
+    {
+      message: `Set up completions?`,
+      options: shellsWithCompletion.map((s) => {
```

Comment:
So this is an interactive prompt - are we sure we want to prompt user and require more interaction? I'm fine either way.

Reply:
For now I'd prefer to have it opt-in, since setting up completions is more invasive that adding to path, and more dependent on your shell setup.

We can change it in the future if it's too annoying